bin/pacignore

```diff
@@ -0,0 +1,238 @@
+  fi
+}
+
+rm_ignorepkg() {
+  # Declare local variables
+  local pkg="$1"
+  local new_array=()
+  local ignorepkg
+
+  # Remove package from array if it exists
+  if check_ignorepkg "$pkg"; then
+    for ignorepkg in "${PACIGNORE_IGNOREPKG[@]}"; do
+      if [ ! "$ignorepkg" == "$pkg" ]; then
+        new_array+=("$ignorepkg")
+      fi
+    done
+
+    # Replace global array with local one
+    PACIGNORE_IGNOREPKG=("${new_array[@]}")
+  else
+    {
+      eval_gettext "Skipping \$pkg as it is was never ignored"
+      printf "\n"
+    } >&2
+    return 1
+  fi
+}
+
+main() {
+  # Declare local variable
+  local pkg exit_code=0 success=0
+  local pacignore_ignorepkg
+
+  # Proceed conditionally given previous exit code
+  if ! pacignore_ignorepkg="$(pacman-conf --config="$PACMAN_CONF" "IgnorePkg" 2>/dev/null)"; then
+    exit_code=1
+    {
+      gettext "Error in reading pacman configuration file"
+      printf "\n"
+    } >&2
+  else
+    # Convert string to array
+    mapfile -t "PACIGNORE_IGNOREPKG" < <(printf "%s" "$pacignore_ignorepkg")
+
+    # Proceed by pacignore subcommands
+    if [ "$PACIGNORE_SUBCOMMAND" == "ls" ]; then
+      printf "%s\n" "${PACIGNORE_IGNOREPKG[@]}"
+    else
+      for pkg; do
+        if ! "${PACIGNORE_SUBCOMMAND}_ignorepkg" "$pkg"; then
+          exit_code=1
+        else
+          success=1
+        fi
+      done
+
+      # Write to pacman configuration file if there is success
+      if ((success)) && [ "$PACIGNORE_SUBCOMMAND" != "check" ]; then
+        write_ignorepkg
+      fi
+    fi
+  fi
+
+  # Return consolidated exit code
+  return "$exit_code"
+}
+
+parse_options() {
+  # Parse the subcommand correctly
+  if [[ "$1" =~ ^(ls|check|add|rm)$ ]]; then
+    PACIGNORE_SUBCOMMAND="$1"
+    shift
+  fi
+
+  # Loop across all CLI arguments and categorize
+  while [[ -n "$1" ]]; do
+    case "$1" in
+      --pacman-conf)
+        if [[ -n "$2" ]]; then
+          shift
+          PACMAN_CONF="$1"
+        else
+          {
+            gettext "Missing --pacman-conf argument"
+            printf "\n"
+            usage
+          } >&2
+          return 1
+        fi
+        ;;
+      -*)
+        local current_option
+        # shellcheck disable=SC2034
+        current_option="$1"
+        {
+          eval_gettext "Unrecognized option \$current_option"
+          printf "\n"
+          usage
+        } >&2
+        return 1
+        ;;
+      *)
+        pkgs=("$@")
+        break
+        ;;
+    esac
+    shift
+  done
+
+  # Check if any packages provided
+  if ((!"${#pkgs[@]}")) && [ "$PACIGNORE_SUBCOMMAND" != "ls" ]; then
+    {
+      gettext "No packages provided"
+      printf "\n"
+      usage
+    } >&2
+    return 1
+  fi
+}
+
+cli() {
+  # Parse all CLI options
+  parse_options "$@"
+
+  # Proceed with main workflow
+  main "${pkgs[@]}"
+}
+
+# Set script defaults
+PACMAN_CONF="/etc/pacman.conf"
+PACIGNORE_SUBCOMMAND="ls"
+
+# CLI execution (not for testing)
+if ((!LIB)); then
```

Comment:
Part of the appeal extracting separate executables, is you now have a smaller unit you can test -- without using `source` and then calling internal functions. `pacignore` is pretty simple, and I think could be unit-tested directly without this machinery.

```
It lists the thing,

    $ echo "whatever" > something.conf
    > PACMAN_CONF=something.conf pacignore ls
    whatever
``` 

Know what I mean?

IMO, the fact that `downgrade` is tested the way it is is a bit of an anti-pattern, so I wouldn't want to spread it to another script like this.

Reply:
Got it and changed this. Now `pacignore` is tested on a script-level